File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.js
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(target, nums) {
    let right = 0, left = 0, winSum = 0, minSize = nums.length + 1;
    
    while(right < nums.length && 1 < minSize){
        winSum += nums[right];
        
        if(winSum >= target){
            while(target <= winSum - nums[left]){
                winSum -= nums[left++];
            }
            minSize = Math.min(minSize, right - left + 1);
            winSum -= nums[left++];
        }
        right++;
    }
    return minSize < nums.length + 1 ? minSize : 0;
};

File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.java
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int minLen = Integer.MAX_VALUE;
        int sum = 0;
        int start = 0;
        
        for(int end = 0; end < nums.length; end++){
            sum += nums[end];
            
            while(sum >= target){
                minLen = Math.min(minLen, end - start + 1);
                sum -= nums[start];
                start++;
            }
        }
        
        return minLen == Integer.MAX_VALUE ? 0 : minLen;
    }
}